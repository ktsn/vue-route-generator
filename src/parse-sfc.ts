import type { parseComponent as _parserV2 } from 'vue-template-compiler'
import type { parse as _parserV3 } from '@vue/compiler-sfc'

export interface ParseResult {
  customBlocks: CustomBlock[]
}

export interface CustomBlock {
  type: string
  content: string
}

export function parseSFC(code: string): ParseResult {
  try {
    const parserV2: typeof _parserV2 = require('vue-template-compiler')
      .parseComponent
    return parserV2(code, {
      pad: 'space',
    })
  } catch {
    try {
      const parserV3: typeof _parserV3 = require('@vue/compiler-sfc').parse
      return parserV3(code, {
        pad: 'space',
      }).descriptor
    } catch {
      throw new Error(
        '[vue-route-generator] Either "vue-template-compiler" or "@vue/compiler-sfc" is required.'
      )
    }
  }
}
