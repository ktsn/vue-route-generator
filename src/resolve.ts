import { NestedMap, setToMap } from './nested-map'
import { parseSFC } from './parse-sfc'

export interface PageMeta {
  name: string
  chunkName: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string
  children?: PageMeta[]
  routeMeta?: any
  route?: any
}

interface FileError extends Error {
  file?: string
}

const routeMetaName = 'route-meta'
const routeBlockName = 'route'

export function resolveRoutePaths(
  paths: string[],
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string
): PageMeta[] {
  const map: NestedMap<string[]> = {}

  const splitted = paths.map((p) => p.split('/'))
  splitted.forEach((path) => {
    setToMap(map, pathToMapPath(path), path)
  })

  return pathMapToMeta(map, importPrefix, nested, readFile)
}

function pathMapToMeta(
  map: NestedMap<string[]>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number = 0
): PageMeta[] {
  if (map.value) {
    const path = map.value

    const meta: PageMeta = {
      name: pathToName(path),
      chunkName: pathToChunkName(path),
      specifier: pathToSpecifier(path),
      path: pathToRoute(path, parentDepth, nested),
      pathSegments: toActualPath(path),
      component: importPrefix + path.join('/'),
    }

    const content = readFile(path.join('/'))
    const parsed = parseSFC(content)
    const routeMetaBlock = parsed.customBlocks.find(
      (b) => b.type === routeMetaName
    )
    const routeBlock = parsed.customBlocks.find(
      (b) => b.type === routeBlockName
    )

    // Deprecated. Will be removed in a later version
    if (routeMetaBlock) {
      console.warn(
        '<route-meta> custom block is deprecated. Use <route> block instead. Found in ' +
          path.join('/')
      )
      meta.routeMeta = tryParseCustomBlock(
        routeMetaBlock.content,
        path,
        'route-meta'
      )
    }

    if (routeBlock) {
      meta.route = tryParseCustomBlock(routeBlock.content, path, 'route')
    }

    if (map.children) {
      meta.children = pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        meta.pathSegments.length
      )
    }

    return [meta]
  }

  return map.children
    ? pathMapChildrenToMeta(
        map.children,
        importPrefix,
        nested,
        readFile,
        parentDepth
      )
    : []
}

function routePathComparator(a: string[], b: string[]): number {
  const a0 = a[0]
  const b0 = b[0]

  if (!a0 || !b0) {
    return a.length - b.length
  }

  const aOrder = isDynamicRoute(a0) ? 1 : 0
  const bOrder = isDynamicRoute(b0) ? 1 : 0
  const order = aOrder - bOrder

  return order !== 0 ? order : routePathComparator(a.slice(1), b.slice(1))
}

function pathMapChildrenToMeta(
  children: Map<string, NestedMap<string[]>>,
  importPrefix: string,
  nested: boolean,
  readFile: (path: string) => string,
  parentDepth: number
): PageMeta[] {
  return Array.from(children.values())
    .reduce<PageMeta[]>((acc, value) => {
      return acc.concat(
        pathMapToMeta(value, importPrefix, nested, readFile, parentDepth)
      )
    }, [])
    .sort((a, b) => {
      // Prioritize static routes than dynamic routes
      return routePathComparator(a.pathSegments, b.pathSegments)
    })
}

function tryParseCustomBlock(
  content: string,
  filePath: string[],
  blockName: string
): any {
  try {
    return JSON.parse(content)
  } catch (err) {
    const joinedPath = filePath.join('/')
    const wrapped: FileError = new Error(
      `Invalid json format of <${blockName}> content in ${joinedPath}\n` +
        err.message
    )

    // Store file path to provide useful information to downstream tools
    // like friendly-errors-webpack-plugin
    wrapped.file = joinedPath

    throw wrapped
  }
}

function isDynamicRoute(segment: string): boolean {
  return segment[0] === ':'
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.vue` from the last path
 * - Omit if the last segument is `index`
 * - Convert dynamic route to `:param` format
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])
  segments = segments.slice(0, -1).concat(last)

  return segments
    .map((s, i) => {
      if (s[0] === '_') {
        const suffix = lastIndex === i ? '?' : ''
        return ':' + s.slice(1) + suffix
      } else {
        return s
      }
    })
    .filter((s) => !isOmittable(s))
}

function pathToMapPath(segments: string[]): string[] {
  const last = segments[segments.length - 1]
  return segments.slice(0, -1).concat(basename(last))
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments
    .slice(0, -1)
    .concat(basename(last))
    .filter((s) => !isOmittable(s))

  if (segments.length === 0) {
    return 'index'
  }

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToChunkName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments.slice(0, -1).concat(basename(last))

  return segments
    .map((s) => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const last = segments[segments.length - 1]
  const replaced = segments
    .slice(0, -1)
    .concat(basename(last))
    .join('_')
    .replace(/[^a-zA-Z0-9]/g, '_')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return prefix + toActualPath(segments).slice(parentDepth).join('/')
}

function basename(filename: string): string {
  return filename.replace(/\.[^.]+$/g, '')
}
