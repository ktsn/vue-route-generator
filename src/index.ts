import * as fs from 'fs'
import * as path from 'path'
import * as fg from 'fast-glob'
import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'

export interface GenerateConfig {
  pages: string
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
  inlineRouteBlock?: boolean
}

export function generateRoutes({
  pages,
  importPrefix = '@/pages/',
  dynamicImport = true,
  chunkNamePrefix = '',
  nested = false,
  inlineRouteBlock = true,
}: GenerateConfig): string {
  const patterns = ['**/*.vue', '!**/__*__.vue', '!**/__*__/**']

  const pagePaths = fg.sync(patterns, {
    cwd: pages,
    onlyFiles: true,
  })

  const metaList = resolveRoutePaths(pagePaths, {
    importPrefix,
    nested,
    inlineRouteBlock,
    readFile: (file) => {
      return fs.readFileSync(path.join(pages, file), 'utf8')
    },
  })

  return createRoutes(
    metaList,
    dynamicImport,
    chunkNamePrefix,
    inlineRouteBlock
  )
}
