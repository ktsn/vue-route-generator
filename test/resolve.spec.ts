import { resolveRoutePaths } from '../src/resolve'

function mockReadFile(path: string): string {
  if (path === 'meta.vue') {
    return `<route-meta>
    {
      "title": "Hello"
    }
    </route-meta>`
  }

  if (path === 'invalid-meta.vue') {
    return `<route-meta>
    {
      "invalid": "Test",
    }
    </route-meta>`
  }

  if (path === 'route.vue') {
    return `<route>
    {
      "name": "Test",
      "meta": {
        "title": "Hello"
      }
    }
    </route>`
  }

  if (path === 'invalid-route.vue') {
    return `<route>
    {
      "name": "Test",
    }
    </route>`
  }

  return ''
}

describe('Route resolution', () => {
  interface TestOptions {
    nested?: boolean
    expectWarn?: string
  }

  function test(
    name: string,
    paths: string[],
    options: TestOptions = {}
  ): void {
    it(name, () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation()

      expect(
        resolveRoutePaths(paths, '@/pages/', !!options.nested, mockReadFile)
      ).toMatchSnapshot()

      if (options.expectWarn) {
        expect(spy).toHaveBeenCalledWith(options.expectWarn)
      }
    })
  }

  test('resolves routes', ['index.vue', 'foo.vue', 'baz.vue'])

  test('resolves nested route including index', ['a/index/b/index/c.vue'])

  test('resolves nested route including index with corresponding index.vue', [
    'a/index.vue',
    'a/index/b/index.vue',
    'a/index/b/index/c.vue',
  ])

  test('resolves dynamic routes', ['users/_id.vue'])

  test('resolves dynamic routes with multiple "."', ['user.register.vue'])

  test('resolves dynamic routes with required param', ['users/_id/index.vue'])

  test('resolves optional dynamic route under index', ['index/_param.vue'])

  test('resolves nested routes', ['foo.vue', 'foo/index.vue', 'foo/bar.vue'])

  test('resolves dynamic nested routes', [
    'users/_id.vue',
    'users/test.vue',
    'users.vue',
    'users/_id/foo.vue',
  ])

  test('resolves spase nested routes', ['users.vue', 'users/session/login.vue'])

  test('resolves number prefixed route', ['1test.vue', '1test/2nested.vue'])

  test('resolve route meta', ['meta.vue'], {
    expectWarn:
      '<route-meta> custom block is deprecated. Use <route> block instead. Found in meta.vue',
  })

  test('resolve route custom block', ['route.vue'])

  test('resolves as nested routes', ['index.vue', 'foo.vue'], {
    nested: true,
  })

  test('prioritizes index than dynamic route', [
    'users/_id.vue',
    'users/foo.vue',
    'users/index.vue',
  ])

  it('throws error when failed to parse route-meta', () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation()

    expect(() => {
      resolveRoutePaths(['invalid-meta.vue'], '@/pages/', false, mockReadFile)
    }).toThrow(
      /Invalid json format of <route-meta> content in invalid-meta\.vue/
    )
    expect(spy).toHaveBeenCalledWith(
      '<route-meta> custom block is deprecated. Use <route> block instead. Found in invalid-meta.vue'
    )
  })

  it('throws error when failed to parse route custom block', () => {
    expect(() => {
      resolveRoutePaths(['invalid-route.vue'], '@/pages/', false, mockReadFile)
    }).toThrow(/Invalid json format of <route> content in invalid-route\.vue/)
  })

  describe('sorting', () => {
    test('prioritizes static routes than dynamic ones', [
      'nested/foo.vue',
      'nested/_id.vue',
      'nested/bar.vue',
    ])

    test('prioritizes deeper routes', [
      'nested/_id/foo/bar.vue',
      'nested/_id/_key/bar.vue',
      'nested/test/foo/bar.vue',
      'nested/test/foo/_id.vue',
    ])

    test('handles when a dynamic route is a directory', [
      'nested/_id/foo.vue',
      'nested/foo.vue',
    ])

    test('handles when a static route is a directory', [
      'nested/_id.vue',
      'nested/foo/bar.vue',
    ])

    test('handles when both static and dynamic routes are directories', [
      'nested/_id/foo.vue',
      'nested/static/foo.vue',
    ])
  })
})
