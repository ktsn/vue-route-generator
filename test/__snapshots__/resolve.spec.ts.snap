// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route resolution prioritizes index than dynamic route 1`] = `
Array [
  Object {
    "chunkName": "users-index",
    "component": "@/pages/users/index.vue",
    "name": "users",
    "path": "/users",
    "pathSegments": Array [
      "users",
    ],
    "specifier": "users_index",
  },
  Object {
    "chunkName": "users-foo",
    "component": "@/pages/users/foo.vue",
    "name": "users-foo",
    "path": "/users/foo",
    "pathSegments": Array [
      "users",
      "foo",
    ],
    "specifier": "users_foo",
  },
  Object {
    "chunkName": "users-id",
    "component": "@/pages/users/_id.vue",
    "name": "users-id",
    "path": "/users/:id?",
    "pathSegments": Array [
      "users",
      ":id?",
    ],
    "specifier": "users__id",
  },
]
`;

exports[`Route resolution resolve route custom block 1`] = `
Array [
  Object {
    "chunkName": "route",
    "component": "@/pages/route.vue",
    "name": "route",
    "path": "/route",
    "pathSegments": Array [
      "route",
    ],
    "route": Object {
      "meta": Object {
        "title": "Hello",
      },
      "name": "Test",
    },
    "specifier": "route",
  },
]
`;

exports[`Route resolution resolve route meta 1`] = `
Array [
  Object {
    "chunkName": "meta",
    "component": "@/pages/meta.vue",
    "name": "meta",
    "path": "/meta",
    "pathSegments": Array [
      "meta",
    ],
    "routeMeta": Object {
      "title": "Hello",
    },
    "specifier": "meta",
  },
]
`;

exports[`Route resolution resolves as nested routes 1`] = `
Array [
  Object {
    "chunkName": "index",
    "component": "@/pages/index.vue",
    "name": "index",
    "path": "",
    "pathSegments": Array [],
    "specifier": "index",
  },
  Object {
    "chunkName": "foo",
    "component": "@/pages/foo.vue",
    "name": "foo",
    "path": "foo",
    "pathSegments": Array [
      "foo",
    ],
    "specifier": "foo",
  },
]
`;

exports[`Route resolution resolves dynamic nested routes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "chunkName": "users-test",
        "component": "@/pages/users/test.vue",
        "name": "users-test",
        "path": "test",
        "pathSegments": Array [
          "users",
          "test",
        ],
        "specifier": "users_test",
      },
      Object {
        "children": Array [
          Object {
            "chunkName": "users-id-foo",
            "component": "@/pages/users/_id/foo.vue",
            "name": "users-id-foo",
            "path": "foo",
            "pathSegments": Array [
              "users",
              ":id",
              "foo",
            ],
            "specifier": "users__id_foo",
          },
        ],
        "chunkName": "users-id",
        "component": "@/pages/users/_id.vue",
        "name": "users-id",
        "path": ":id?",
        "pathSegments": Array [
          "users",
          ":id?",
        ],
        "specifier": "users__id",
      },
    ],
    "chunkName": "users",
    "component": "@/pages/users.vue",
    "name": "users",
    "path": "/users",
    "pathSegments": Array [
      "users",
    ],
    "specifier": "users",
  },
]
`;

exports[`Route resolution resolves dynamic routes 1`] = `
Array [
  Object {
    "chunkName": "users-id",
    "component": "@/pages/users/_id.vue",
    "name": "users-id",
    "path": "/users/:id?",
    "pathSegments": Array [
      "users",
      ":id?",
    ],
    "specifier": "users__id",
  },
]
`;

exports[`Route resolution resolves dynamic routes with multiple "." 1`] = `
Array [
  Object {
    "chunkName": "user.register",
    "component": "@/pages/user.register.vue",
    "name": "user.register",
    "path": "/user.register",
    "pathSegments": Array [
      "user.register",
    ],
    "specifier": "user_register",
  },
]
`;

exports[`Route resolution resolves dynamic routes with required param 1`] = `
Array [
  Object {
    "chunkName": "users-id-index",
    "component": "@/pages/users/_id/index.vue",
    "name": "users-id",
    "path": "/users/:id",
    "pathSegments": Array [
      "users",
      ":id",
    ],
    "specifier": "users__id_index",
  },
]
`;

exports[`Route resolution resolves nested route including index 1`] = `
Array [
  Object {
    "chunkName": "a-index-b-index-c",
    "component": "@/pages/a/index/b/index/c.vue",
    "name": "a-b-c",
    "path": "/a/b/c",
    "pathSegments": Array [
      "a",
      "b",
      "c",
    ],
    "specifier": "a_index_b_index_c",
  },
]
`;

exports[`Route resolution resolves nested route including index with corresponding index.vue 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "chunkName": "a-index-b-index-c",
            "component": "@/pages/a/index/b/index/c.vue",
            "name": "a-b-c",
            "path": "c",
            "pathSegments": Array [
              "a",
              "b",
              "c",
            ],
            "specifier": "a_index_b_index_c",
          },
        ],
        "chunkName": "a-index-b-index",
        "component": "@/pages/a/index/b/index.vue",
        "name": "a-b",
        "path": "b",
        "pathSegments": Array [
          "a",
          "b",
        ],
        "specifier": "a_index_b_index",
      },
    ],
    "chunkName": "a-index",
    "component": "@/pages/a/index.vue",
    "name": "a",
    "path": "/a",
    "pathSegments": Array [
      "a",
    ],
    "specifier": "a_index",
  },
]
`;

exports[`Route resolution resolves nested routes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "chunkName": "foo-index",
        "component": "@/pages/foo/index.vue",
        "name": "foo",
        "path": "",
        "pathSegments": Array [
          "foo",
        ],
        "specifier": "foo_index",
      },
      Object {
        "chunkName": "foo-bar",
        "component": "@/pages/foo/bar.vue",
        "name": "foo-bar",
        "path": "bar",
        "pathSegments": Array [
          "foo",
          "bar",
        ],
        "specifier": "foo_bar",
      },
    ],
    "chunkName": "foo",
    "component": "@/pages/foo.vue",
    "name": "foo",
    "path": "/foo",
    "pathSegments": Array [
      "foo",
    ],
    "specifier": "foo",
  },
]
`;

exports[`Route resolution resolves number prefixed route 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "chunkName": "1test-2nested",
        "component": "@/pages/1test/2nested.vue",
        "name": "1test-2nested",
        "path": "2nested",
        "pathSegments": Array [
          "1test",
          "2nested",
        ],
        "specifier": "_1test_2nested",
      },
    ],
    "chunkName": "1test",
    "component": "@/pages/1test.vue",
    "name": "1test",
    "path": "/1test",
    "pathSegments": Array [
      "1test",
    ],
    "specifier": "_1test",
  },
]
`;

exports[`Route resolution resolves optional dynamic route under index 1`] = `
Array [
  Object {
    "chunkName": "index-param",
    "component": "@/pages/index/_param.vue",
    "name": "param",
    "path": "/:param?",
    "pathSegments": Array [
      ":param?",
    ],
    "specifier": "index__param",
  },
]
`;

exports[`Route resolution resolves routes 1`] = `
Array [
  Object {
    "chunkName": "index",
    "component": "@/pages/index.vue",
    "name": "index",
    "path": "/",
    "pathSegments": Array [],
    "specifier": "index",
  },
  Object {
    "chunkName": "foo",
    "component": "@/pages/foo.vue",
    "name": "foo",
    "path": "/foo",
    "pathSegments": Array [
      "foo",
    ],
    "specifier": "foo",
  },
  Object {
    "chunkName": "baz",
    "component": "@/pages/baz.vue",
    "name": "baz",
    "path": "/baz",
    "pathSegments": Array [
      "baz",
    ],
    "specifier": "baz",
  },
]
`;

exports[`Route resolution resolves spase nested routes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "chunkName": "users-session-login",
        "component": "@/pages/users/session/login.vue",
        "name": "users-session-login",
        "path": "session/login",
        "pathSegments": Array [
          "users",
          "session",
          "login",
        ],
        "specifier": "users_session_login",
      },
    ],
    "chunkName": "users",
    "component": "@/pages/users.vue",
    "name": "users",
    "path": "/users",
    "pathSegments": Array [
      "users",
    ],
    "specifier": "users",
  },
]
`;

exports[`Route resolution sorting handles when a dynamic route is a directory 1`] = `
Array [
  Object {
    "chunkName": "nested-foo",
    "component": "@/pages/nested/foo.vue",
    "name": "nested-foo",
    "path": "/nested/foo",
    "pathSegments": Array [
      "nested",
      "foo",
    ],
    "specifier": "nested_foo",
  },
  Object {
    "chunkName": "nested-id-foo",
    "component": "@/pages/nested/_id/foo.vue",
    "name": "nested-id-foo",
    "path": "/nested/:id/foo",
    "pathSegments": Array [
      "nested",
      ":id",
      "foo",
    ],
    "specifier": "nested__id_foo",
  },
]
`;

exports[`Route resolution sorting handles when a static route is a directory 1`] = `
Array [
  Object {
    "chunkName": "nested-foo-bar",
    "component": "@/pages/nested/foo/bar.vue",
    "name": "nested-foo-bar",
    "path": "/nested/foo/bar",
    "pathSegments": Array [
      "nested",
      "foo",
      "bar",
    ],
    "specifier": "nested_foo_bar",
  },
  Object {
    "chunkName": "nested-id",
    "component": "@/pages/nested/_id.vue",
    "name": "nested-id",
    "path": "/nested/:id?",
    "pathSegments": Array [
      "nested",
      ":id?",
    ],
    "specifier": "nested__id",
  },
]
`;

exports[`Route resolution sorting handles when both static and dynamic routes are directories 1`] = `
Array [
  Object {
    "chunkName": "nested-static-foo",
    "component": "@/pages/nested/static/foo.vue",
    "name": "nested-static-foo",
    "path": "/nested/static/foo",
    "pathSegments": Array [
      "nested",
      "static",
      "foo",
    ],
    "specifier": "nested_static_foo",
  },
  Object {
    "chunkName": "nested-id-foo",
    "component": "@/pages/nested/_id/foo.vue",
    "name": "nested-id-foo",
    "path": "/nested/:id/foo",
    "pathSegments": Array [
      "nested",
      ":id",
      "foo",
    ],
    "specifier": "nested__id_foo",
  },
]
`;

exports[`Route resolution sorting prioritizes deeper routes 1`] = `
Array [
  Object {
    "chunkName": "nested-test-foo-bar",
    "component": "@/pages/nested/test/foo/bar.vue",
    "name": "nested-test-foo-bar",
    "path": "/nested/test/foo/bar",
    "pathSegments": Array [
      "nested",
      "test",
      "foo",
      "bar",
    ],
    "specifier": "nested_test_foo_bar",
  },
  Object {
    "chunkName": "nested-test-foo-id",
    "component": "@/pages/nested/test/foo/_id.vue",
    "name": "nested-test-foo-id",
    "path": "/nested/test/foo/:id?",
    "pathSegments": Array [
      "nested",
      "test",
      "foo",
      ":id?",
    ],
    "specifier": "nested_test_foo__id",
  },
  Object {
    "chunkName": "nested-id-foo-bar",
    "component": "@/pages/nested/_id/foo/bar.vue",
    "name": "nested-id-foo-bar",
    "path": "/nested/:id/foo/bar",
    "pathSegments": Array [
      "nested",
      ":id",
      "foo",
      "bar",
    ],
    "specifier": "nested__id_foo_bar",
  },
  Object {
    "chunkName": "nested-id-key-bar",
    "component": "@/pages/nested/_id/_key/bar.vue",
    "name": "nested-id-key-bar",
    "path": "/nested/:id/:key/bar",
    "pathSegments": Array [
      "nested",
      ":id",
      ":key",
      "bar",
    ],
    "specifier": "nested__id__key_bar",
  },
]
`;

exports[`Route resolution sorting prioritizes static routes than dynamic ones 1`] = `
Array [
  Object {
    "chunkName": "nested-foo",
    "component": "@/pages/nested/foo.vue",
    "name": "nested-foo",
    "path": "/nested/foo",
    "pathSegments": Array [
      "nested",
      "foo",
    ],
    "specifier": "nested_foo",
  },
  Object {
    "chunkName": "nested-bar",
    "component": "@/pages/nested/bar.vue",
    "name": "nested-bar",
    "path": "/nested/bar",
    "pathSegments": Array [
      "nested",
      "bar",
    ],
    "specifier": "nested_bar",
  },
  Object {
    "chunkName": "nested-id",
    "component": "@/pages/nested/_id.vue",
    "name": "nested-id",
    "path": "/nested/:id?",
    "pathSegments": Array [
      "nested",
      ":id?",
    ],
    "specifier": "nested__id",
  },
]
`;
